### PR TITLE
Add basic support for Zookeeper Serversets for SD

### PR DESCRIPTION
**Pull request by brian-brazil**
It can discover an entire tree of serversets, or just one.
This is a very naieve approach using polling,
this should be changed to use watches in the future.

For anyone reading this in future: This is not how you're
meant to use Zookeeper. Watches are far more efficient.

@fabxc

### CODE REVIEW

**Comment by fabxc** [comment]
I understand serversets are a defined schemes for doing SD in ZK, which is great as it gives us something we can support.

Several concerns here:
- I'm not familiar with serversets. Does this cover the whole topic of serversets? Are they always JSON – if not which other encodings? If there's anything serverset SD could be extended with, will this break the current config?
- Google shows very few results for the whole serverset thing. It does not seem like something very common to me. I'd like to see more people expressing demand for it.
- `this should be changed to use watches in the future.` – by whom? Once this hits master it will always be at the bottom of someones backlog.
- The implementation does not implement target provider interface, it circumvents it.

Each discovery mechanism comes with an overhead (config, dependencies, code). We have to establish some requirements for native support. Those should be:

- Mechanism is a legitimate way of doing SD [✓]
- Sufficient demand [✗]
- Sound implementation [✗] that interacts with our target manager as intended [✗]

Confd + FileSD yields better results here.

**Comment by brian-brazil** [comment]
> I'm not familiar with serversets. Does this cover the whole topic of serversets? Are they always JSON – if not which other encodings? If there's anything serverset SD could be extended with, will this break the current config?

They can also be Thrift, but it seems most people use JSON.

> Google shows very few results for the whole serverset thing. It does not seem like something very common to me. I'd like to see more people expressing demand for it.

It's common for those using the Twitter stack, such as Finagle and Aurora. I've already had a few people interested in it.

> The implementation does not implement target provider interface, it circumvents it.

Ah, I forgot to finish off that bit.

**Comment by brian-brazil** [comment]
Okay, it took some work but this is using watches now.

**Comment by brian-brazil** [comment]
That should be good now.

File: retrieval/discovery/serverset.go [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+			fmt.Sprintf("%d", endpoint.Port))
+
+	}
+
+	labels[serversetStatusLabel] = clientmodel.LabelValue(member.Status)
+
+	return &labels, nil
+}
+
+type ZookeeperTreeCache struct {
```
Comment (fabxc) [inline retrieval/discovery/serverset.go]: This one and the one below can be unexported, right?

File: retrieval/discovery/serverset.go [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+	conn.SetLogger(ZookeeperLogger{})
+	if err != nil {
+		return nil
+	}
+	updates := make(chan ZookeeperTreeCacheEvent, 20000)
+	sd := &ServersetDiscovery{
+		conf:      conf,
+		conn:      conn,
+		updates:   updates,
+		sources:   make(map[string]struct{}),
```
Comment (fabxc) [inline retrieval/discovery/serverset.go]: We have the "no `make()` unless necessary" style guideline, so `map[string]struct{}{}`.

File: retrieval/discovery/serverset.go (hunk 1 above)
Comment (brian-brazil) [inline retrieval/discovery/serverset.go]: Yes, I'm planning on offering this upstream as an initial Go version of https://curator.apache.org/apidocs/org/apache/curator/framework/recipes/cache/TreeCache.html so I was setting visibility accordingly.

File: retrieval/discovery/serverset.go [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+	children, _, childWatcher, err := tc.conn.ChildrenW(path)
+	if err == zk.ErrNoNode {
+		tc.recursiveDelete(path, node)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	currentChildren := make(map[string]struct{})
```
Comment (fabxc) [inline retrieval/discovery/serverset.go]: `map[string]struct{}{}`

File: retrieval/discovery/serverset.go [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+	for _, child := range children {
+		currentChildren[child] = struct{}{}
+		childNode := node.children[child]
+		// Does not already exists, create it.
+		if childNode == nil {
+			node.children[child] = &zookeeperTreeCacheNode{
+				events:   node.events,
+				children: make(map[string]*zookeeperTreeCacheNode),
```
Comment (fabxc) [inline retrieval/discovery/serverset.go]: dito

File: retrieval/discovery/serverset.go [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+			} else {
+				path := strings.TrimPrefix(ev.Path, tc.prefix)
+				parts := strings.Split(path, "/")
+				node := tc.head
+				for _, part := range parts[1:] {
+					childNode := node.children[part]
+					if childNode == nil {
+						childNode = &zookeeperTreeCacheNode{
+							events:   tc.head.events,
+							children: make(map[string]*zookeeperTreeCacheNode),
```
Comment (fabxc) [inline retrieval/discovery/serverset.go]: dito

File: retrieval/discovery/serverset.go [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+	AdditionalEndpoints map[string]serversetEndpoint
+	Status              string `json:"status"`
+}
+
+type serversetEndpoint struct {
+	Host string
+	Port int
+}
+
+type ZookeeperLogger struct {
```
Comment (fabxc) [inline retrieval/discovery/serverset.go]: Can be removed with the next vendor upgrade of our log package.

**Comment by fabxc** [comment]
We manually re-insert LICENSE files and such into the vendoring directory – not entirely sure why the NOTICE file doesn't suffice legally either.

**Comment by fabxc** [comment]
Wow, I didn't expect it to be this complicated since it boiled down to [this](https://github.com/fabxc/prom_sd_example/blob/master/etcd/main.go#L96-L202) for etcd.
So thanks a ton for the effort!

Aside from minor comments :+1: for the code – obviously haven't tested it myself.

**Comment by beorn7** [comment]
@fabxc The details about the license inclusion recursively depends on the license itself. In doubt, we must err on the side of including it.

**Comment by brian-brazil** [comment]
I believe I've addressed all comments, and reworked the run/sources handling.

> Wow, I didn't expect it to be this complicated

Why do you think I started with a polling approach :)

**Comment by fabxc** [comment]
Great :+1:

**Pull request merged**